acct/update-email

```diff
@@ -34,111 +26,102 @@ there is a problem with your account.
 Your email address is kept private, and can only be viewed by OCF staff. Please
 enter an email address below that you can be reached at.
 
-For more information, see http://ocf.io/email-update
-================================================================================"""
-
-
-def email_str(maybe_email):
-    """Returns whether maybe_email could be an email address."""
-    return EMAIL_REGEX.match(maybe_email) is not None
+For more information, see https://ocf.io/email-update
+================================================================================
+"""
 
 
-def valid_email(maybe_email):
-    """Returns whether maybe_email would be a valid email address. It must
-    represent a possible email, and not be an @ocf email address."""
+def valid_non_ocf_email(email):
+    """Returns whether something is a valid email address. It must
+    represent a valid email, and not be an @ocf email address."""
 
-    if not maybe_email or not email_str(maybe_email):
+    if not email or not valid_email(email):
         return False
 
-    # @ocf emails are no longer supported.
+    # @ocf emails are no longer supported
     #
     # This can be subverted by modifying LDAP directly, but it's a reasonable
     # sanity check.
-    if maybe_email.lower().endswith('@ocf.berkeley.edu'):
+    if email.lower().endswith('@ocf.berkeley.edu'):
         print('You may not use an @ocf.berkeley.edu email address.')
         return False
 
     return True
 
 
-def get_ldap():
-    """Returns LDAP connection, or None."""
-    try:
-        auth = ldap.sasl.gssapi('')
-        conn = ldap.initialize(LDAP_SERVER)
-        conn.sasl_interactive_bind_s('', auth)
-        return conn
-    except ldap.LOCAL_ERROR:
-        return None
-
-
-def get_email(conn, dn):
+def get_email(username):
     """Returns current email, or None."""
-    results = conn.search_s(dn, ldap.SCOPE_SUBTREE, attrlist=[MAIL_ATTR])
 
-    if results:
-        attrs = results[0][1]
-        return attrs.get(MAIL_ATTR, [None])[0]
+    # Since the mail attribute is private, and we can't get the attribute's
+    # value without authenticating, we have to use ldapsearch here instead of
+    # something like ldap3.
+    output = subprocess.check_output(
+        ('ldapsearch', '-LLL', 'uid={}'.format(username), LDAP_MAIL_ATTR),
+        stderr=subprocess.DEVNULL,
+    ).decode('utf-8').split('\n')
 
+    mail_attr = [attr for attr in output if attr.startswith(LDAP_MAIL_ATTR + ': ')]
 
-def set_email(conn, dn, old_email, new_email):
-    """Updates email attribute in LDAP."""
-    old = {MAIL_ATTR: '*'}
-    new = {MAIL_ATTR: new_email}
+    if mail_attr:
+        # Strip the '{LDAP_MAIL_ATTR}: ' from the beginning of the string
+        return mail_attr[0][len(LDAP_MAIL_ATTR) + 2:].strip()
```

Comment:
Do we care to check that there is only one mail attribute?

Reply:
If there was, what would we do? Return both? Error out? It seems to me like just returning the first one is a pretty sane behavior.

I ended up modifying the LDAP config (stored in `cn=config cn={0}core`) and changed the `mail` attribute so that it can only have a single value (now it's just like `loginShell` in that behavior), and I think that essentially resolves this issue. I only found one account that had two `mail` attributes, and fixed that by using the one that was listed on their website as their contact address. I did a search using this command:

```
ldapsearch -LLL uid=* mail | pcregrep -B 2 -M "mail: .+\nmail:"
```